[Cases] Custom fields MVP

js-jankisalvi opened the pull request:
## Summary

This PR implements an MPV for custom fields in Cases. Specifically:

- Users can add, delete, or edit custom fields from the configuration page
- Users can set custom fields when creating a case
- Users can set custom fields when editing a case
- Only the `Text` custom field type is supported. The `Toggle` is implemented to test the framework.
- Users with a basic license can enter the configuration page and set custom fields
- The configuration page header changed to "Settings"
- The "Edit external connection" button changed to "Settings"
- APIs:
    - Users cannot post custom fields with duplicate keys
    - Users cannot change the type of the custom field
    - Users cannot add custom fields that are not configured
    - Required fields should be present
    - If some of the custom fields are omitted from the request the backend will fill them with `null` values 
- Limits:
    - Maximum 10 custom fields configured
    - Maximum 50 chars in custom field labels
    - Maximum 160 chars for the value of the `Text` custom field type
    - Users cannot change the type of a custom field
    - The key of the custom fields should contain only lowercase letters (a-z), numbers (0-9), '_', and '-'
    - Required fields need a value

## Testing
- Cases created before custom fields are working as expected (create some on `main` before switching to the feature branch)
- Environments without configuration are working as expected (clear ES data)
- Environments with existing configurations are working as expected
- Try to create cases with custom fields (required & optional) and delete some of them
- Try to create cases with custom fields (required & optional) and then configure new custom fields. Existing cases with old custom fields should work as expected
- User actions are working as expected especially with the combinations described in the previous two bullets
- Users with a basic license can configure custom fields but cannot configure connectors
- Users with a gold+ license can configure custom fields and connectors
- Users with read access cannot configure or edit custom fields

Resolves: https://github.com/elastic/kibana/issues/160236

PRs:

- https://github.com/elastic/kibana/pull/165671
- https://github.com/elastic/kibana/pull/166353
- https://github.com/elastic/kibana/pull/166439
- https://github.com/elastic/kibana/pull/166483
- https://github.com/elastic/kibana/pull/166815
- https://github.com/elastic/kibana/pull/166940
- https://github.com/elastic/kibana/pull/166962
- https://github.com/elastic/kibana/pull/166969
- https://github.com/elastic/kibana/pull/166975
- https://github.com/elastic/kibana/pull/167029
- https://github.com/elastic/kibana/pull/167047
- https://github.com/elastic/kibana/pull/167105
- https://github.com/elastic/kibana/pull/167131
- https://github.com/elastic/kibana/pull/167144
- https://github.com/elastic/kibana/pull/167166
- https://github.com/elastic/kibana/pull/167167
- https://github.com/elastic/kibana/pull/167310
- https://github.com/elastic/kibana/pull/167386
- https://github.com/elastic/kibana/pull/167481
- https://github.com/elastic/kibana/pull/167495
- https://github.com/elastic/kibana/pull/167398
- https://github.com/elastic/kibana/pull/167472

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses sentence case text and includes [i18n support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Documentation](https://www.elastic.co/guide/en/kibana/master/development-documentation.html) was added for features that require explanation or tutorials
- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios

### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

## Release notes
Allow users to configure custom fields in Cases. Supported field types: Text and Toggle. Coming soon: List, Number, Date, IP, Email, etc.

adcoelho committed:
Cases mapping changes.

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into cases-custom-fields-feature-branch

kibanamachine committed:
[CI] Auto-commit changed files from 'node scripts/check_mappings_update --fix'

adcoelho committed:
Merge remote-tracking branch 'upstream/main' into cases-custom-fields-feature-branch

adcoelho committed:
[Cases] Custom fields types (#165671)

Connected with #160236

**Merging into a Feature Branch**

## Summary

This PR defines the `customField` in the Cases API and Domain types.

For now, we only allow the custom fields `string`, `boolean`, and
`list`.

This PR does not include the case configuration types.

The default value for the field is `[]`.

adcoelho committed:
[Cases] Custom Fields API documentation (#166353)

## Summary

This was missing in my previous PR #165671

adcoelho committed:
[Cases] Add customFields to cases configuration (#166439)

Following up on @js-jankisalvi work I changed the `customFields`
property in `CasesConfigure` to be the most basic possible.

Changes:
- We only accept `TextCustomField` and `ToggleCustomField` in the cases
configuration api.
- The custom field type used in the Cases domain and API types now comes
from the enum in `configure/v1`.
- Cases configuration POST requests now require the `customFields`
property

js-jankisalvi committed:
[Cases] added UI to add custom field on configuration page (text and toggle field) (#166483)

## Summary

Connected to https://github.com/elastic/kibana/issues/160236


### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing
guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses
sentence case text and includes [i18n
support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios
- [x] Any UI touched in this PR is usable by keyboard only (learn more
about [keyboard accessibility](https://webaim.org/techniques/keyboard/))
- [x] This renders correctly on smaller devices using a responsive
layout. (You can test this [in your
browser](https://www.browserstack.com/guide/responsive-testing-on-local-server))

### For maintainers

- [x] This was checked for breaking API changes and was [labeled
appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Christos Nasikas <christos.nasikas@elastic.co>

cnasikas committed:
Merge branch 'main' into cases-custom-fields-feature-branch

cnasikas committed:
Merge branch 'cases-custom-fields-feature-branch' of github.com:elastic/kibana into cases-custom-fields-feature-branch

cnasikas committed:
[Cases] Model version the Cases SO for custom fields (#166940)

cnasikas committed:
Merge branch 'main' into cases-custom-fields-feature-branch

cnasikas committed:
[Cases] Edit custom fields in the case view page (#166815)

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Janki Salvi <117571355+js-jankisalvi@users.noreply.github.com>

adcoelho committed:
[Cases] Backend validation custom fields (#166962)

## Summary

This PR implements the following validation:

- **Custom field keys should be unique**
  - Where:
- `CaseConfigureService` -
x-pack/plugins/cases/server/services/configure/index.ts
      - `post`
      - `patch` 
  - Tests:
    - x-pack/plugins/cases/server/services/configure/index.test.ts
- **Limit custom fields to 5**
  - Where:
    - `ConfigurationRequestRt` (POST)
    - `ConfigurationPatchRequestRt`
    - `CasePostRequestRt` - done in a previous PR
    - `CasePatchRequestRt` - done in a previous PR
  - Tests:
    - x-pack/plugins/cases/server/client/configure/client.test.ts
    - x-pack/plugins/cases/common/types/api/configure/v1.test.ts
    - x-pack/plugins/cases/server/client/cases/create.test.ts
    - x-pack/plugins/cases/server/client/cases/update.test.ts

js-jankisalvi committed:
Merge branch 'main' into cases-custom-fields-feature-branch

adcoelho committed:
[Cases] Validate length of text custom fields. (#167029)

## Summary

This PR limits the length of the property value for `text` custom
fields.

adcoelho committed:
[Cases] Validate custom field keys (#166975)

## Summary

This PR implements the following validation:

- **Custom Field Keys need to exist in configuration when creating or
updating cases**
  - Where:
    - Cases Client
      - `create`
      - `update` 
  - Tests:
-
x-pack/test/cases_api_integration/security_and_spaces/tests/common/cases/patch_cases.ts
-
x-pack/test/cases_api_integration/security_and_spaces/tests/common/cases/post_case.ts

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

cnasikas committed:
[Cases] Adds the experimental badge to custom fields (#167105)

js-jankisalvi committed:
[Cases] Create case form with custom fields (#166969)

## Summary

This PR adds the ability to add custom fields in the create case form. 

**Testing**

Verify that you all custom fields from configuration are visible in
create case form
Verify that you can set your custom fields
Verify that it throws an error based on the configuration (required:
true | false)

Connected to https://github.com/elastic/kibana/issues/160236

**Create case**


https://github.com/elastic/kibana/assets/117571355/8b560804-ea1f-4313-a382-34a17c0750b7


### Checklist

- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: Christos Nasikas <christos.nasikas@elastic.co>

js-jankisalvi committed:
[Cases] Limit number of custom fields in UI (#167047)

## Summary

Connected to https://github.com/elastic/kibana/issues/160236

This PR adds validation in UI to limit maximum number of custom fields
to 5 in configuration cases page


![image](https://github.com/elastic/kibana/assets/117571355/7cbb2dfb-bf13-498c-a4a1-1fddeefa4d46)

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing
guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses
sentence case text and includes [i18n
support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

cnasikas committed:
Merge branch 'main' into cases-custom-fields-feature-branch

adcoelho committed:
[Cases] Required custom fields in the create case API (#167131)

## Summary

This PR checks for missing required custom fields in the create case
API.

(In my next PR I am thinking of moving `casesClient.configure.get({
owner: casePostRequest.owner });` outside of the validation functions.)

js-jankisalvi committed:
[Cases] Delete custom field (#167167)

## Summary

This PR allows users to delete custom fields in settings page.


https://github.com/elastic/kibana/assets/117571355/c8387543-c09b-4c4f-84f7-97708fa12006

### Checklist

Delete any items that are not applicable to this PR.

- [x] Any text added follows [EUI's writing
guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses
sentence case text and includes [i18n
support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

cnasikas committed:
Merge branch 'main' into cases-custom-fields-feature-branch

adcoelho committed:
[Cases] Validate types custom fields (#167166)

Co-authored-by: Christos Nasikas <christos.nasikas@elastic.co>

cnasikas committed:
[Cases] Custom fields user actions (#167144)

js-jankisalvi committed:
[Cases] Edit custom field configuration (#167386)

## Summary

This PR allows users to edit custom field configuration.


https://github.com/elastic/kibana/assets/117571355/efaf9a32-b0e2-4d72-89e9-2045c544af2b

connected to https://github.com/elastic/kibana/issues/160236
### Checklist

- [x] Any text added follows [EUI's writing
guidelines](https://elastic.github.io/eui/#/guidelines/writing), uses
sentence case text and includes [i18n
support](https://github.com/elastic/kibana/blob/main/packages/kbn-i18n/README.md)
- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

cnasikas committed:
[Cases] Custom fields schema changes (#167310)

cnasikas committed:
Merge branch 'main' into cases-custom-fields-feature-branch

adcoelho committed:
[Cases] Change max custom fields per case (#167481)

## Summary

Change max custom fields per Case limit to 10.

Connected to https://github.com/elastic/kibana/issues/160236

adcoelho committed:
[Cases] Prevent patching custom field types (#167495)

## Summary

In this PR we prevent users from sending PATCH requests to the cases
configuration that change the type of a custom field.

Connected to https://github.com/elastic/kibana/issues/160236

jcger committed:
[Cases] Settings UI License Limitations (#167398)

### Description
In the prior implementation, users lacking a gold license were unable to
access the configure/settings page since all available options were
exclusively tied to the gold license. However, with this update, the
capability to incorporate custom actions into this page is extended to
basic license holders as well. This PR accomplishes this by hiding
sections that demand gold-level privileges from basic license users.

#### Now with Basic License:
<img width="897" alt="Screenshot 33"
src="https://github.com/elastic/kibana/assets/17549662/d9ec7d96-db77-4833-a7c7-e90617280b14">

#### Before with Basic License:
The Cases page was initially inaccessible due to the disabled Settings
button. However, it remained accessible via the URL, which displayed the
same page as it did for Gold users.
<img width="900" alt="Screenshot"
src="https://github.com/elastic/kibana/assets/17549662/be53227d-0089-4b7e-97a2-5f8033eec621">

### QA
#### Test that Basic License users only see Custom Fields section
1. Go to Cases > Settings
2. You shouldn't see any section but the Custom Fields

#### If you want to test that the callout doesn't show up for Basic
License users:
1. Go to Stack Management > License Management > Start Trial
2. Go to Cases > Settings 
4. Add a new connector in the External incident management system
section
5. Delete that connector in Stack Management > Connectors 
6. Check that in Cases > Settings a waning callout is shown now
7. Go to Stack Management > License Management > Revert to Basic
8. Check that in Cases > Settings the warning callout is not shown 

Connected to https://github.com/elastic/kibana/issues/160236

cnasikas committed:
Merge branch 'main' into cases-custom-fields-feature-branch

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

elasticmachine commented:
Pinging @elastic/response-ops-cases (Feature:Cases)

gsoldevila reviewed (approved):
Only additive (thus compatible) changes in the mappings, LGTM!

cnasikas committed:
Fix core test

cnasikas reviewed (commented):
Great job everyone with this feature 🚀. I did a thorough test of the feature including the UI and the APIs and I found the following bugs. 

- UI: Cannot create a case that has a toggle and text fields. I get an error because the UI sets the value of the toggle to `value: [value]` where it should be `value: value`
- UI: Does not validate 160 chars if the field is optional
- UI: Label validation message says "fieldLabel" should change to "field label"
- API: Can create a case without custom fields when the configuration has required fields

js-jankisalvi committed:
[Cases] Custom fields feature fix tests (#167472)

## Summary

Connected to https://github.com/elastic/kibana/issues/160236

This PR fixes tests for custom field mvp feature

### Checklist

- [x] [Unit or functional
tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html)
were updated or added to match the most common scenarios

---------

Co-authored-by: Christos Nasikas <christos.nasikas@elastic.co>
Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>
Co-authored-by: adcoelho <antonio.coelho@elastic.co>

banderror reviewed (approved):
The change in `x-pack/test/security_solution_cypress/cypress/objects/case.ts` 👍

js-jankisalvi committed:
fix UI bugs

js-jankisalvi committed:
Merge branch 'cases-custom-fields-feature-branch' of https://github.com/elastic/kibana into cases-custom-fields-feature-branch

js-jankisalvi commented:
> * UI: Cannot create a case that has a toggle and text fields. I get an error because the UI sets the value of the toggle to `value: [value]` where it should be `value: value`
> * UI: Does not validate 160 chars if the field is optional
> * UI: Label validation message says "fieldLabel" should change to "field label"

fixed in https://github.com/elastic/kibana/pull/167016/commits/787c040e8bfd6014b13286961354b713a7160626

cnasikas committed:
Fix test dependencies

cnasikas commented on x-pack/plugins/cases/public/components/create/form_context.tsx:
```diff
@@ -105,10 +105,16 @@ export const FormContext: React.FC<Props> = ({
 
         if (configCustomField) {
-          const fieldValue = isEmpty(value) && typeof value !== 'boolean' ? null : [value];
+          if (!isEmpty(value) && typeof value === 'string') {
```
Let's put this in a helper function so we can easily extend it when we add more field types. Also, let's add some unit tests (if we don't have any) to test that boolean and string values are working as expected.

cnasikas replied:
Nice!

cnasikas committed:
Fix current_mappings.json

cnasikas commented:
> Great job everyone with this feature 🚀. I did a thorough test of the feature including the UI and the APIs and I found the following bugs.
> 
> * UI: Cannot create a case that has a toggle and text fields. I get an error because the UI sets the value of the toggle to `value: [value]` where it should be `value: value`
> * UI: Does not validate 160 chars if the field is optional
> * UI: Label validation message says "fieldLabel" should change to "field label"
> * API: Can create a case without custom fields when the configuration has required fields

I tested all the bugs found and everything is working as expected!

jcger committed:
throw if request is sending an unknown custom field

js-jankisalvi committed:
helper function for custom field value conversion

cnasikas commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into cases-custom-fields-feature-branch

cnasikas commented on x-pack/plugins/cases/server/client/cases/validators.test.ts:
```diff
@@ -0,0 +1,436 @@
+      ).toThrowErrorMatchingInlineSnapshot(`"Missing required custom fields: first_key"`);
+    });
+
+    it('throws if request sends a custom field that is not part of the configuration', () => {
```
I am wondering if this check exceeds the responsibility of the validation function which is to validate if required fields are present. The `validateCustomFieldKeysAgainstConfiguration` validates for fields that are not part of the configuration. Though they are not connected in this context and are tested in isolation (test in the create/update routes should ensure that both the validation functions are being called). Wdyt?

jcger replied:
you are right, I'll remove it from the validateRequiredCustomFields fn

cnasikas commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into cases-custom-fields-feature-branch

cnasikas commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into cases-custom-fields-feature-branch

cnasikas commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into cases-custom-fields-feature-branch

jcger committed:
remove notRegisteredCustomFields - duplicated

cnasikas committed:
Fix cypress cases connectors tests

cnasikas committed:
Merge branch 'cases-custom-fields-feature-branch' of github.com:elastic/kibana into cases-custom-fields-feature-branch

cnasikas commented:
buildkite test this

cnasikas commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into cases-custom-fields-feature-branch

cnasikas commented:
Blocked by https://github.com/elastic/kibana/pull/167745

cnasikas commented:
@elasticmachine merge upstream

kibanamachine committed:
Merge branch 'main' into cases-custom-fields-feature-branch

cnasikas reviewed (approved):
@js-jankisalvi @adcoelho @jcger Great work 🚀
